### DISCUSSION TITLE
Error: Can't combine an optional file parameter with a json data parameter

### DISCUSSION THREAD

**Discussion by Ousman1301** [fastapi/fastapi#12763] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, HTTPException, File, UploadFile, Request, Response
from docx import Document
from docx.shared import Pt, RGBColor
from docx.enum.text import WD_PARAGRAPH_ALIGNMENT
from docx.oxml import OxmlElement
from docx.oxml.ns import qn
from io import BytesIO
from typing import List, Dict, Optional
import json
import re
from docx2pdf import convert
import tempfile
import os


@app.post("/insert-table-in-word-test/", tags=["WORD"])
def insert_table_in_a_word_document(
        data: Dict[str, str],
        # Accept data as a dictionary from the request body, row_italic: bool = False, row_color: Optional[str] = None,
        table_borders: bool = True, table_alignment: str = "left",file: Optional[bytes] = File(None)
):

    doc = Document()
    buf = BytesIO()
    doc.save(buf)
    buf.seek(0)

    # Return the Word file
    return Response(
        content=buf.read(),
        media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        headers={"Content-Disposition": "attachment; filename=generated_table.docx"}
    )


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

I have created an endpoint where I simply want to add tables in a word file. This endpoint works perfectly fine without the file parameter Optional[bytes] = File(None). I have used that parameter earlier in an other endpoint, but it seems like the combination of a json parameter (data) and Optional[bytes] = File(None) is not compatible. What's also surprising with this is that I am not even using the file parameter in my endpoint.

As mentioned - if you try this code without the File parameter, this endpoint will work fine.

Also: When I am not specifying a file, the content type is application/json. As soon as I specify a file, it becomes multipart/form-data. Maybe it's something there?

### Operating System

Windows

### FastAPI Version

0.115.2

### Pydantic Version

2.9.2

### Python Version

Python 3.12.3

### Additional Context

I am getting this error message:
{
  "detail": [
    {
      "type": "dict_type",
      "loc": [
        "body",
        "data"
      ],
      "msg": "Input should be a valid dictionary",
      "input": "{\"additionalProp1\":\"string\",\"additionalProp2\":\"string\",\"additionalProp3\":\"string\"}"
    }
  ]
}

**Comment by luzzodev** [comment, marked as answer]
Usually sending json and bytes like objects is not supported by default.
You are trying to send/receive something like `multipart/mixed` that is not supported by default on fastapi and usually neither on browser or  other libs.

There are different approaches you can use to achive the same result:

Use `multipart/form-data`:
You can encode your json as a string and pass it as a form field. The you have to manually decode or parse your json string.

Use `application/json`:
You can encode your file bytes into a string and sent it as a pure json request.

Use a custom implementation:
You can implement something like:  `multipart/mixed`, but you have to manually parse the body and extract all fields and files.

**Comment by Ousman1301** [comment]
I see. I often see application/json as the prefered content type, so perhaps I should stick to that.

How can I specify the file parameter in a way that takes it as a string ?

**Reply by luzzodev** [reply]
to convert bytes in to a string you can encode file bytes in base64 and decode in utf-8.
to convert json into a string you can just dump it.

All this convertions must be done on client side.
On server side you should handle converted type and only later convert it back.

Just pay attention that encoding in base64 will increase the size of file of 33% more or less.

So choosing between sending data as multipart/form-data or application/json strictly depends on your implementation requirements and use case.

**Reply by Ousman1301** [reply]
I see! I tried specifying it as application/json, and it worked perfectly. 

Seems like I have to implement this a bit different than intended, but this was very helpful. Thank you so much! <3
